### PR TITLE
Extract the source type into a separate package.

### PR DESCRIPTION
**Pull request by immutableT**
Extract the source type into a separate package.

### CODE REVIEW

File: internal/pkg/source/source_test.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	Doc:        "test harness for the logic related to sources",
+	Requires:   []*analysis.Analyzer{buildssa.Analyzer},
+	ResultType: reflect.TypeOf(analyzerResult{}),
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	in := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA)
+	var result analyzerResult
+	for _, fn := range in.SrcFuncs {
+		fmt.Printf("Func: %v\n", fn.Name())
```
Comment (PurelyApplied) [inline internal/pkg/source/source_test.go:69]: I would prefer these to be `t.Logf` statements so they don't clutter up the command-line `go test ./...`  Although to do that, you'd have to do some closure tricks and reset `testAnalyzer.Run` at the start of every test.

If you want to keep the prints, I'd still prefer the closure.  Otherwise, if it was just for debugging, please remove.

File: internal/pkg/source/source_test.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+		for _, b := range fn.Blocks {
+			for _, i := range b.Instrs {
+				fmt.Printf("%v : %T\n", i, i)
```
Comment (PurelyApplied) [inline internal/pkg/source/source_test.go:72]: See about regarding `t.Logf`.

File: internal/pkg/source/source_test.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		fieldsPattern:      "name",
+	}
+
+	testCases := []struct {
+		desc                         string
+		pattern                      string
+		config                       *testConfig
+		wantCallConnectionIdx        []int
+		wantFieldAccessConnectionIdx []int
+		wantStoreConnectionIdx       []int
```
Comment (PurelyApplied) [inline internal/pkg/source/source_test.go:105]: Testing based on index in the `ssa` result feels to me a bit brittle and resistant to future change.  See above comment about `// want` approach.

File: internal/pkg/source/source_test.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+		},
+		{
+			desc:                         "sanitization",
+			pattern:                      "sanitization",
+			config:                       config,
+			wantFieldAccessConnectionIdx: []int{0},
+		},
+	}
+
+	for _, tt := range testCases {
```
Comment (PurelyApplied) [inline internal/pkg/source/source_test.go:128]: This took me a while to understand, and I think every check done here would be performed by a `// want "allocation"`.

File: internal/pkg/source/testdata/src/sanitization/test.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+import "fmt"
+
+type foo struct {
+	name     string
+	password string
+}
+
+func f1() {
+	foo := &foo{name: "bar", password: "P@ssword01"}
```
Comment (PurelyApplied) [inline internal/pkg/source/testdata/src/sanitization/test.go:25]: Might be best to avoid name-showing unless we're explicitly testing for it.

File: internal/pkg/utils/utils_test.go [hunk 6]
```diff
@@ -27,7 +27,7 @@ import (
 )
 
 var testAnalyzer = &analysis.Analyzer{
-	Name:       "domination",
+	Name:       "dereference",
```
Comment (PurelyApplied) [inline internal/pkg/utils/utils_test.go:30]: This rename is outside the scope of the commit message.  Please either update the commit message or separate this to another PR.

File: internal/levee.go [hunk 7]
```diff
@@ -544,11 +467,11 @@ func readConfig(path string) (*config, error) {
 	return readConfigCached, readConfigCachedErr
 }
 
-func identifySources(conf *config, ssaInput *buildssa.SSA) map[*ssa.Function][]*source {
-	sourceMap := make(map[*ssa.Function][]*source)
+func identifySources(conf *config, ssaInput *buildssa.SSA) map[*ssa.Function][]*source.Source {
```
[the remark this answers is not in the document]
Comment (immutableT) [reply internal/levee.go:470]: I think so.
In this PR I wanted to focus on the logic of the Source type.

File: internal/pkg/source/source.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+// Source.sanitized notes sanitizer calls that sanitize this Source
+type Source struct {
+	node       ssa.Node
+	marked     map[ssa.Node]bool
+	sanitizers []*sanitizer.Sanitizer
+	config     classifier
+}
+
+// Node returns the underlying ssa.Node of the Source.
+func (a *Source) Node() ssa.Node {
```
[the remark this answers is not in the document]
Comment (immutableT) [reply internal/pkg/source/source.go:45]: I agree. 
I will fix this on a separate PR.

[the remark this answers is not in the document]
Comment (immutableT) [reply internal/pkg/source/source.go:45]: In general, I agree.
However, since a Source could be referenced by Values and Instructions, it forces us to use ssa.Node to keep track of vertices, right?

File: internal/pkg/source/source_test.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	var result analyzerResult
+	for _, fn := range in.SrcFuncs {
+		fmt.Printf("Func: %v\n", fn.Name())
+		for _, b := range fn.Blocks {
+			for _, i := range b.Instrs {
+				fmt.Printf("%v : %T\n", i, i)
+
+				switch v := i.(type) {
+				case *ssa.Alloc:
+					result.allocations = append(result.allocations, v)
```
[the remark this answers is not in the document]
Comment (immutableT) [reply internal/pkg/source/source_test.go:76]: I would certainly prefer the approach of the `analysistest`.
However, this would imply creating a new analyzer for each test case, right?
The statement above, assumes that we want to test more than just identifying the source, but also the relationships between the source and its referrers.

File: internal/pkg/utils/utils_test.go (hunk 6 above)
Comment (immutableT) [reply internal/pkg/utils/utils_test.go]: Reverted.

File: internal/pkg/source/source_test.go (hunk 2 above)
Comment (immutableT) [reply internal/pkg/source/source_test.go]: Removed print statements.

File: internal/pkg/source/source_test.go (hunk 3 above)
Comment (immutableT) [reply internal/pkg/source/source_test.go:101]: See my previous [response](https://github.com/google/go-flow-levee/pull/11#discussion_r424108402).

File: internal/pkg/source/source_test.go (hunk 4 above)
Comment (immutableT) [reply internal/pkg/source/source_test.go:124]: These tests go beyond just identifying where allocation is performed for a Source.
Concretely, these tests ensure that we:
1. do not build the graph of a Source beyond a sanitizer
2. do include calls to propagators in the graph

File: internal/pkg/source/testdata/src/sanitization/test.go (hunk 5 above)
Comment (immutableT) [reply internal/pkg/source/testdata/src/sanitization/test.go]: Replaced with simpler values.

File: internal/pkg/source/source_test.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+	dir := analysistest.TestData()
+	config := &testConfig{
+		propagatorsPattern: "propagator",
+		sanitizerPattern:   "sanitizer",
+		fieldsPattern:      "name",
+	}
+
+	testCases := []struct {
+		desc                         string
+		pattern                      string
```
[the remark this answers is not in the document]
Comment (immutableT) [reply internal/pkg/source/source_test.go]: Removed the `desc` field.

File: internal/levee.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (PurelyApplied) [reply internal/levee.go:470]: Fair enough.

File: internal/pkg/source/source.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (PurelyApplied) [reply internal/pkg/source/source.go:45]: I'm not sure that's the case.  Generally, a source will be sanitize/propagated/sunk by an `ssa.Instruction`

The following are the only `ssa` types which implement `Value` but not `Instruction`.

```
*Builtin
*Const
*FreeVar
*Function
*Global
*Parameter
```

So long as those types are part of source identification, I don't think we'll lose coverage.  But it would require another pass at the `bfs` / dominance logic, so probably shouldn't be part of this PR.

File: internal/pkg/source/source_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (PurelyApplied) [reply internal/pkg/source/source_test.go:72]: > However, this would imply creating a new analyzer for each test case, right?
Ahh, that's a fair point.  Depending on how you split things, it might not be every test-case, but could have to be one analyzer for one or more `testdata/` files.

**Comment by PurelyApplied** [comment]
As an aside, I would appreciate it if you didn't force push on your branch after the PR is open.  It makes it a harder to track changes across comments.  Since I've pulled your branch to look at the diff locally, I now have to blow the branch away for future examination rather than just pull new changes.  We can squash before merge if necessary.

File: internal/pkg/source/source.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (immutableT) [reply internal/pkg/source/source.go:45]: I think we are on the same page.
I agree that the constructor of the source should accept `ssa.Value` and not `ssa.Node`. 

However, the graph of the source will contain Instructions (example `ssa.Store`), and since the `Source` itself is part of the graph, vertices would need to support a broader interface - probably Node.
Let's pick this up on a separate PR.

File: internal/pkg/source/source_test.go (hunk 1 above)
Comment (immutableT) [reply internal/pkg/source/source_test.go]: Removed.

File: internal/pkg/source/source_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (immutableT) [reply internal/pkg/source/source_test.go:72]: The fundamental problem is that ssa instructions do not translate one-to-one to the original Go source code. 
Concretely, if I wanted to use the `analysistest` approach and assert that a given source has the expected graph, I would not be able to do so via comments alone since not all ssa instructions have an equivalent in the original source.
Therefore, a different approach is required. However, I hope that we could improve on the approach suggested by the PR by developing some helper types that reduce the cognitive overhead.

File: internal/pkg/source/testdata/src/allocation/test.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"fmt"
+)
+
+type foo struct {
+	name string
+}
+
+func f1() {
+	foo := &foo{name: "bar"}
```
Comment (PurelyApplied) [inline internal/pkg/source/testdata/src/allocation/test.go:26]: Also name shadowed.

File: internal/pkg/source/testdata/src/propagation/test.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+package propagation
+
+import "fmt"
+
+type foo struct {
+	name string
+}
+
+func f1() {
+	foo := &foo{name: "bar"}
```
Comment (PurelyApplied) [inline internal/pkg/source/testdata/src/propagation/test.go:24]: Also name shadowed.

File: internal/pkg/source/testdata/src/sanitization/test.go [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+import "fmt"
+
+type foo struct {
+	name     string
+	password string
+}
+
+func f1() {
+	foo := &foo{name: "n", password: "p"}
```
Comment (PurelyApplied) [inline internal/pkg/source/testdata/src/sanitization/test.go:25]: Also name shadowed.

File: internal/pkg/source/source.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (PurelyApplied) [reply internal/pkg/source/source.go:45]: > Let's pick this up on a separate PR.

+1.
